packages/node_modules/@webex/internal-plugin-encryption/src/kms-certificate-validation.js

```diff
@@ -0,0 +1,229 @@
+          break;
+        }
+      }
+    }
+  }
+
+  if (!validationSuccessful) {
+    // Didn't match kid in the Subject Alt Names, checking the Common Name
+    const subjectAttributes = certificate.subject.typesAndValues;
+
+    for (const attribute of subjectAttributes) {
+      if (attribute.type === X509_COMMON_NAME_KEY) {
+        const commonName = attribute.value.valueBlock.value;
+
+        validationSuccessful = commonName === kidHostname;
+        if (validationSuccessful) {
+          break;
+        }
+      }
+    }
+  }
+
+  if (!validationSuccessful) {
+    throwError('hostname of the 1st certificate does not match \'kid\'');
+  }
+};
+
+/**
+ * Validate the first KMS certificate against the information
+ * provided in the JWT
+ * @param {Certificate} certificate first certificate the identifies the KMS
+ * @param {Object} JWT credentials of the KMS
+ * @param {string} JWT.e Public exponent of the first certificate
+ * @param {string} KWT.n Modulus of the first certificate
+ * @throws if e or n doesn't match the first certificate
+ * @returns {null}
+ */
+const validatePublicCertificate =
+  ([certificate], {e: publicExponent, n: modulus}) => {
+    const {encode} = jose.util.base64url;
+
+    const publicKey = certificate.subjectPublicKeyInfo.subjectPublicKey;
+    const asn1PublicCert = fromBER(publicKey.valueBlock.valueHex);
+    const publicCert = new RSAPublicKey({schema: asn1PublicCert.result});
+    const publicExponentHex = publicCert.publicExponent.valueBlock.valueHex;
+    const modulusHex = publicCert.modulus.valueBlock.valueHex;
+
+    if (publicExponent !== encode(publicExponentHex)) {
+      throwError('Public exponent is invalid');
+    }
+    if (modulus !== encode(modulusHex)) {
+      throwError('Modulus is invalid');
+    }
+  };
+
+/**
+ * Validates the list of certificates against the CAs provided
+ * @param {certificate[]} certificates list of certificates provided
+ *   by the KMS to certify itself
+ * @param {string[]} caroots list of Certificate Authorities used to
+ *   validate the KMS's certificates
+ * @returns {Promise} rejects if unable to validate the certificates
+ */
+const validateCertificatesSignature = (certificates, caroots = []) => {
+  const certificateEngine = new CertificateChainValidationEngine({
+    trustedCerts: caroots.map(decodeCert),
+    certs: certificates
+  });
+
+  return certificateEngine.verify()
+    .then(({result, resultCode, resultMessage}) => {
+      if (!result) {
+        throwError(
+          `Certificate Validation failed [${resultCode}]: ${resultMessage}`
+        );
+      }
+    });
+};
+
+/**
+ * Validates the information provided by the KMS
+ * @param {string[]} caroots PEM encoded certificates that will be used
+ *   as Certificate Authorities
+ * @param {Object} jwt Object containing the fields necessary to
+ *   validate the KMS
+ * @returns {Object} jwt
+ */
+const validateKMS = (caroots) => (jwt = {}) => Promise.resolve()
```

Comment:
It's a curried function. The first function takes an array of CA strings and returns another function. The second function takes the credentials of the KMS and validates it.

Reply:
Ah, gotcha, I haven't actually used a curried function outside of a few js exercises. What benefit does this give us over just having the `caroots` be a regular param?

Also, looks like a lot of debate over how the jsdoc should appear: https://github.com/jsdoc/jsdoc/issues/1286#issuecomment-341911947

Would be helpful to at least mention the currying in the description, because otherwise intellisense is going to show this as a function with two params `validateKMS(caroots, jwt)`.

Reply:
Having it as a curried function is similar to having a class. The first function would be the instantiation of the class. It takes constructor variables. In this case, the CA root certificates. 

Keeping the analogy going, we can use this class instance to verify KMS credentials. The second function would be the class method that verifies the KMS credentials.